### DISCUSSION TITLE
The cadquery engine is broken for me, any help?

### DISCUSSION THREAD

**Discussion by MirandaStreeter** [joshreve/dactyl-keyboard#48] [General]
Heya. I was running into previous issues with 6x6 `.step` files not producing key holes for the keyboard. So, I decided to start afresh and eliminate any mistakes I made. I deleted the local repo and wiped docker to get a new blank slate:

```
docker stop $(docker ps -a -q)
docker rm $(docker ps -a -q)
docker system prune --all --volumes --force
systemctl restart docker
```

Then, with a fresh repo, I ran `./dactyl.sh`, and started the interactive `DM-shell` session. I can run `generate_configuration.py` followed by `dactyl_manuform.py` and it spits out the correct `.scad` files in the `/app/things` directory. Cool, works so far!

From there, all I did was switch the engine from `solid` to `cadquery`, and attempted an otherwise default build in order to make sure it works. It doesn't, unfortunately. I'm getting the following traceback:

```
(base) root@1b34df009558:/app/src# python dactyl_manuform.py 
NO CONFIGURATION SPECIFIED, USING run_config.json
Found Current Engine in Config = cadquery
Traceback (most recent call last):
  File "/app/src/dactyl_manuform.py", line 69, in <module>
    from helpers_cadquery import *
  File "/app/src/helpers_cadquery.py", line 1, in <module>
    import cadquery as cq
  File "/opt/conda/lib/python3.9/site-packages/cadquery/__init__.py", line 3, in <module>
    from .occ_impl.shapes import (
  File "/opt/conda/lib/python3.9/site-packages/cadquery/occ_impl/shapes.py", line 20, in <module>
    from vtk import (
  File "/opt/conda/lib/python3.9/site-packages/vtk.py", line 30, in <module>
    all_m = importlib.import_module('vtkmodules.all')
  File "/opt/conda/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "/opt/conda/lib/python3.9/site-packages/vtkmodules/all.py", line 54, in <module>
    from .vtkIONetCDF import *
ImportError: /opt/conda/lib/python3.9/site-packages/vtkmodules/../../.././libnetcdf.so.19: undefined symbol: H5Pset_fapl_ros3
(base) root@1b34df009558:/app/src#
```

Looks like something wasn't linked at runtime. Was there a recent update to the base image that includes a now-incompatible library? Did I miss a step? I'd dig deeper, but I want to at least note my reproduction steps before I get back to this.

Maybe eventually I'll get that 6x6 OpenCascade print going. :)

**Comment by joshreve** [comment]
Someone else also contacted me about this as well.  I likely need to set a fixed python or package version.  I'm guessing something was updated and broke cadquery.  I'm planning to look into it, likely this weekend, but maybe sooner.  I'll comment when it is corrected and commited.

**Reply by joshreve** [reply]
I can't reproduce the error and I'm trying to figure out what is different on my build.  I am on python 3.9.6 and am using the Dockerfile to build, though I am running the build_docker.bat file to create my containers since I am in windows. 

Are you running from a windows or linux host?  If windows, try using the build_docker.bat file.  If Linux, let me know and I can try to get a VM to use docker to see if I can reproduce.

**Reply by joshreve** [reply]
I have replicated the error in a VM.  I can now look at fixing it.

**Reply by joshreve** [reply]
This is apparently not a unique problem [https://github.com/h5py/h5py/issues/1880](https://github.com/h5py/h5py/issues/1880).  I am trying to dictate python=3.8 to see if this fixes it.  I'll push the change if it does.

**Reply by joshreve** [reply]
No obvious solution still.  It is a problem with one of the packages (h5py / hdf5) not calling out the right requirements.  I'm going to try a few other things, but I'm not sure there is a clean solution until they fix their repos.  I'll give one more update on my attempts tomorrow.

**Comment by joshreve** [comment]
I found the solution.  I am modifying the Dockerfile accordingly until the other repo is fixed.  I should be able to push it tonight, there is one more thing in the setup that I have to make sure is automated correctly.

**Reply by joshreve** [reply]
Fixed and pushed to both dev and master.  That was a challenge, but I now have a more strictly defined environment so I should be able to minimize the probability of future issues like this.

**Reply by MirandaStreeter** [reply]
Thanks! Unfortunately, there may have been a regression. Now I'm onto a new error:

```
[INFO] Building docker container: DM-shell...

e12f8ce714f30aa2ce74b1e6e23f9167f94717342482add0a3c84e8e2f06c623

Would you like to start a shell session? [y/n]y

[INFO] Starting session in container: DM-shell

        Type "exit" to terminate the session.

(myenv) root@e12f8ce714f3:/app/src# ls
dactyl_manuform.py  generate_configuration.py  helpers_blender.py  helpers_cadquery.py  helpers_solid.py  model_builder.py  parts  run_config.json
(myenv) root@e12f8ce714f3:/app/src# python generate_configuration.py 
(myenv) root@e12f8ce714f3:/app/src# python dactyl_manuform.py 
NO CONFIGURATION SPECIFIED, USING run_config.json
Found Current Engine in Config = solid
Traceback (most recent call last):
  File "/app/src/dactyl_manuform.py", line 71, in <module>
    from helpers_solid import *
  File "/app/src/helpers_solid.py", line 1, in <module>
    import solid as sl
ModuleNotFoundError: No module named 'solid'
(myenv) root@e12f8ce714f3:/app/src#
```

This is on linux using the `DM-shell` image.

**Reply by joshreve** [reply]
Lol, fortunately this is an easy fix.  I must have accidentally deleted the pip3 install or put out in the wrong spot while fixing cadquery.  I should have it fixed tonight.  In the mean time you should be able to `pip3 install solidpython` in the shell to get it running.

**Reply by joshreve** [reply]
Fixed and pushed, hopefully for good this time!  You may need to prune all you images/containers, but I double checked both engines are running.

**Comment by MirandaStreeter** [comment]
Perfect, thank you! The engine works!

Unfortunately, I'm still running into issues creating a 6x6. The first issue I ran into that led me down the road of rebuilding the container.

Steps I took to reproduce:

* Start `DM-shell` session
* Edit `/app/src/generate_configuration.py`
  * Line 13: Comment out `'ENGINE': 'solid',`
  * Line 14: Uncomment `'# ENGINE': 'cadquery',`
  * Line 27: Set `'nrows':  6`
* Run `python generate_configuration.py`
* Run `python dactyl_manuform.py`
* Open `DM_right.step` in freecad

I then get the following:

_Left_
![image](https://user-images.githubusercontent.com/4233669/138196976-cda960a2-7ada-4e09-bff6-132fd1052d16.png)

_Right_
![image](https://user-images.githubusercontent.com/4233669/138197048-e505009b-e9b7-4d90-85e3-b1757e356c72.png)

And if I change the following (which is more what I'm trying to go for):
  * Line 24: Set `'show_caps': False,`
  * Line 51: Set `'thumb_style': 'DEFAULT',`
  * Line 181: Set `'plate_style': 'HOLE',`
  * Line 218: Set `'oled_mount_type':  'NONE',`


I get the following:

_Left_
![image](https://user-images.githubusercontent.com/4233669/138197781-6b884777-ce36-4774-bda0-dfb9ac7da521.png)

_Right_
![image](https://user-images.githubusercontent.com/4233669/138197664-1709bdcf-93d9-492b-b434-2a1459f1a1ce.png)


Any advice? Is this not a supported config? Am I just bad at freecad? Or should I be using something else to export `.step` to `.stl`? Any help would be greatly appreciated.

Considering this is a separate issue, I can submit a separate discussion if you wish.

**Reply by joshreve** [reply]
Ugh, I'll check it out.  I think something I did for the separable thumb is have secondary effects.  I saw this early on and thought I fixed it.  I also think I am trying to support too many workflows and I am seeing different issues in linux and windows with docker and direct library install.  Either way, I will run through linux / docker to replicate and see.  I think it will be a quick fix.  

Also, I appreciate you communicating these issues.  I don't have the bandwidth to test all the different configs and people communicating when they have an issue really helps me keep the repo operational.

**Reply by joshreve** [reply]
And now I am really confused.  It appears to be a linux vs windows issue.  When I run in docker it is like you are showing, when I run directly in python on windows it is correct.  I will take a look at the order of unions and see if that is the issue.  Perhaps I can create a more robust order that is less susceptible to error in the cadquery library.

**Reply by joshreve** [reply]
So, for clarification it is absolutely a unioning issue, I just find it odd that the libraries for linux and windows act so differently.  I have to look into this.

**Reply by joshreve** [reply]
I think I fixed it and pushed it.  Try it out.  It looks like it works for me now.  Hopefully something new didn't break.

**Reply by MirandaStreeter** [reply]
Unfortunately, I'm seeing the following with the plate the instant I remove the trackball, 6 nrows or not:
![image](https://user-images.githubusercontent.com/4233669/138583362-aa9299f0-21ee-4a0e-ad82-b9bdf5cae769.png)

Also, with or without changes, I'm seeing this on the the keyboard itself (bottom perspective view):
![image](https://user-images.githubusercontent.com/4233669/138583456-b98fb1e9-77f3-400a-b52f-991d99906b5a.png)

This goes for both left and right step files.

**Reply by joshreve** [reply]
I'll take another look.  I really need to consider how to make unit tests, unfortunately it is fairly complex when checking against geometry.

**Reply by joshreve** [reply]
Alright, I think I finally fixed it all . . . I say for the 4th time.  It turns out this was all caused by the cadquery version.  I am surprised how different the versions acted and will have to consider how to carefully control it moving forward.  You should be all set now.  

The newest issues were artifacts from correcting for the wrong version of cadquery.  I reverted those changes and corrected the environment.  You will need to recreate the image and containers again.  I'm really hoping that fixes the geometry differences between windows and linux for good.

**Reply by MirandaStreeter** [reply]
It works! Thanks!

**Reply by joshreve** [reply]
I appreciate the patience and multiple rounds of feedback.  It really helped me stabilize this issue that has been popping up off and on for a while.
